### DISCUSSION TITLE
Padding to Multiples

### DISCUSSION THREAD

**Discussion by exedore-de** [ImageMagick/ImageMagick#6023] [Help] [answered]
### ImageMagick version

7.X

### Operating system, version and so on

Windows 10

### Description

Greetings!

I'm dealing with a bit of a convoluted workflow at a client's request and trying to figure out the best way to create an IM command line/batch under Windows for my needs.

The workflow goes like this:
- Images exported from InDesign (so far, so good)
- Images need to be converted to 1520px wide (so far, so good, can do this already in IM) 
- The height needs to then be padded _to the nearest multiple of 2000px_ (ideally using the majority colour from the bottom row, typically white) - this is the hard part.
- Stitch the resulting images together (this part I've also worked out)

Because the source images are of varying heights, selecting a fixed height output won't work, and the width must remain consistent.

Any help would be appreciated. Thanks a bunch!

**Comment by snibgo** [comment, marked as answer]
> The height needs to then be padded to the nearest multiple of 2000px (ideally using the majority colour from the bottom row, typically white) - this is the hard part.

I assume that:

- If the input height is a multiple of 2000, don't change it.

- Otherwise, increase the height to the nearest multkiple of 2000.

Is that correct? If so, then an expression for the output height is "`ceil(h/2000)*2000`" where `h` is the input height. We can use that in `%[fx:...]` for `-extent`:
```
magick in.png -extent %wx%[fx:ceil(h/2000)*2000] out.png
```
[EDIT: correction from `w` to `h`.]

If you use Windows BAT, then each % needs to be doubled.

On "`-extent`", see https://imagemagick.org/script/command-line-options.php#extent. The padding colour is set by "`-background`". Setting this to the *average* (the mean) of the bottom row is fairly easy. Windows CMD (not BAT) syntax:
```
magick ^
  in.png ^
  ( +clone ^
    -gravity South -crop x1+0+0 +repage -scale "1x1^!" ^
    -set option:COL %[hex:u] ^
    +delete ^
  ) ^
  -background #%[COL] -extent %wx%[fx:ceil(h/2000)*2000] ^
  out.png
```
[EDIT: correction from `w` to `h`.]

"... ideally using the majority colour from the bottom row" is more difficult. What is the "majority colour"? Perhaps the colour that occurs most frequently. If so, then "`histogram:`" will give you that.

**Reply by exedore-de** [reply]
Thank you so much! I was able to hack around the bits I was missing/needed to bring it all together. The final batch for anyone looking in the future:

```
magick ^
  .\PNG\*.png ^
  ( +clone ^
    -gravity South -crop x1+0+0 +repage -scale "1x1^!" ^
    -set option:COL %%[hex:u] ^
    +delete ^
  ) ^
  -resize 1520x -filter lanczos2sharp -units pixelsperinch -density 300x300 -colorspace sRGB^
  -background #%%[COL] -extent %%wx%%[fx:ceil(h/2000)*2000] ^
  out.png

magick convert -append *.png joined.png
del out*.png
```

**Comment by GeeMack** [comment]
This command is very much like the solution **snibgo** offered. It starts by reading in and resizing your input image. Then inside the parentheses it makes a clone, moves the bottom row to the top and peels it off, and resizes that to 1x1 to blend the colors for a new background . After the parentheses it resizes that single pixel to the width of the input image, and to the next height larger than the input height that is evenly divisible by 2000. It finishes with `+swap` to put the input and that newly created background in the right order for compositing, composites the input over that colored background, and writes the output image.

    magick input.png -resize 1520x ^
       ( +clone -roll +0+1 -extent 0x1 -resize 1x1 ) ^
       -resize %[fx:u.w]x%[fx:t?ceil(u.h/2000)*2000:u.h]! ^
       +swap -composite result.png

The input image will be placed at the top on the background. Add a `-gravity` setting of "center" or "south" before the `-composite` to place your input image at the center or bottom of the background.

This is in Windows syntax. For *nix OS change the caret "^" continued-line characters to backslashes "\\", and escape the parentheses and maybe any other special characters with backslashes "\\(...\\)".

**Reply by exedore-de** [reply]
Thanks! This didn't quite pull what I needed it to when trying to batch it up, but it put me on the right track to figuring out the right code order.
